File: maxProduct.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        #Will i repeately see the same problem?
        #describe the state, in this case what set of numbers will define where I am ?
        #set the base cases of answers we know trivially
        #write the rule for the recurrence problem
        #choose engine

        n = len(nums)

        if not nums:
            return 
        res = float('-inf')

        mem = {}
        def product(index):
            nonlocal res
            if index == n-1:
                if nums[index] >= 0:
                    res = max(res, nums[index])
                    return (nums[index],0)
                else:
                    res = max(res, nums[index])
                    return(0, nums[index])
            
            tuple_val = product(index+1)
            if nums[index] >= 0:
                positive = max(nums[index], nums[index]*tuple_val[0])
                negative = min(nums[index], nums[index]*tuple_val[1])
           
            else:
                positive = max(nums[index], nums[index]*tuple_val[1])
                negative = min(nums[index], nums[index]*tuple_val[0])
            
            mem[index] = (positive,negative)
            res = max(res, positive,negative)
            
            return mem[index]

        product(0)
        return res
            

File: maxProduct.rs
impl Solution {
    pub fn max_product(nums: Vec<i32>) -> i32 {
        let mut current_product = nums[0];
        let mut current_product_min = nums[0];
        let mut max_product = nums[0];

        for &num in nums.iter().skip(1){
            if num < 0{
                std::mem::swap(&mut current_product, &mut current_product_min);
            }
            
            current_product = std::cmp::max(num, current_product * num);
            current_product_min = std::cmp::min(num, current_product_min * num);
            max_product = std::cmp::max(current_product, max_product);

        }
        max_product
    }
}
